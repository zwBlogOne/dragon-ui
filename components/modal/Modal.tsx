import React, { Component } from 'react';
import { createPortal } from 'react-dom';
import classnames from 'classnames';
import Events from '../utils/events';
import { ModalProps, StyleType } from './PropsType';

function toggleBodyOverflow(show) {
  let scrollBarWidth = window.innerWidth - document.documentElement.offsetWidth;
  if (show === true) {
    document.body.classList.add('ui-modal-body-overflow');
    if (scrollBarWidth > 0) {
      document.body.style.setProperty('padding-right', scrollBarWidth + 'px');
    }
  } else {
    document.body.classList.remove('ui-modal-body-overflow');
    document.body.style.setProperty('padding-right', null);
  }
}

class Modal extends Component<ModalProps, any> {

  static Header: any;
  static Body: any;
  static Footer: any;

  static defaultProps = {
    prefixCls: 'ui-modal',
    visible: false,
    animationType: 'zoom',
    animationDuration: 300,
    width: 600,
    minWidth: 270,
    isRadius: false,
    isRound: false,
    onMaskClick() { },
  };

  private modal: HTMLDivElement | null;
  private div: HTMLDivElement = document.createElement('div');

  constructor(props) {
    super(props);
    this.state = {
      isShow: false,
      isPending: false,
      animationState: 'leave',
    };
    this.animationEnd = this.animationEnd.bind(this);
  }

  componentWillMount() {
    if (this.props.visible) {
      this.enter();
    }
  }

  componentDidMount() {
    document.body.appendChild(this.div);
  }

  componentWillUpdate() {
    Events.on(this.modal, 'webkitAnimationEnd', this.animationEnd);
    Events.on(this.modal, 'animationend', this.animationEnd);
  }

  componentWillUnmount() {
    Events.off(this.modal, 'webkitAnimationEnd', this.animationEnd);
    Events.off(this.modal, 'animationend', this.animationEnd);
    toggleBodyOverflow(false);
    setTimeout(() => {
      document.body.removeChild(this.div);
    });
  }

  componentWillReceiveProps(nextProps) {
    if (!this.props.visible && nextProps.visible) {
      this.enter();
    } else if (this.props.visible && !nextProps.visible) {
      this.leave();
    }
  }

  shouldComponentUpdate(_, nextState) {
    return !!(this.state.isShow || nextState.isShow);
  }

  animationEnd() {
    if (this.state.animationState === 'leave') {
      this.setState({
        isShow: false,
        isPending: false,
      });
    } else {
      this.setState({
        isShow: true,
        isPending: false,
      });
    }
  }

  enter() {
    toggleBodyOverflow(true);
    this.setState({
      isShow: true,
      isPending: true,
      animationState: 'enter',
    });
  }

  leave() {
    this.setState({
      isShow: true,
      isPending: true,
      animationState: 'leave',
    });
    toggleBodyOverflow(false);
  }

  render() {
    const {
      prefixCls,
      animationType,
      animationDuration,
      width,
      minWidth,
      isRadius,
      isRound,
      className,
      onMaskClick,
      children,
    } = this.props;
    const { isShow, isPending, animationState } = this.state;

    const classes = {
      modal: classnames({
        [prefixCls!]: true,
        radius: 'radius' in this.props || isRadius,
        round: 'round' in this.props || isRound,
        [`fade-${animationState}`]: isPending,
        [className!]: !!className,
      }),
      dialog: classnames({
        [`${prefixCls}-dialog`]: true,
        [`${animationType}-${animationState}`]: true,
      }),
    };

    const style: StyleType = {
      modal: {
        WebkitAnimationDuration: `${animationDuration}ms`,
        MozAnimationDuration: `${animationDuration}ms`,
        msAnimationDuration: `${animationDuration}ms`,
        OAnimationDuration: `${animationDuration}ms`,
        animationDuration: `${animationDuration}ms`,
        position: 'fixed',
      },
      dialog: {
        width: Number(width),
        minWidth: Number(minWidth),
        WebkitAnimationDuration: `${animationDuration}ms`,
        MozAnimationDuration: `${animationDuration}ms`,
        msAnimationDuration: `${animationDuration}ms`,
        OAnimationDuration: `${animationDuration}ms`,
        animationDuration: `${animationDuration}ms`,
      },
    };
    if (!isShow) {
      style.modal.display = 'none';
    }
    return createPortal(
      <div
        className={classes.modal}
        style={style.modal}
        onClick={onMaskClick}
        ref={(ele) => { this.modal = ele; }}
      >
        <div className={`${prefixCls}-wrapper`}>
          <div
            className={classes.dialog}
            style={style.dialog}
            onClick={e => e.stopPropagation()}
          >
            {children}
          </div>
        </div>
      </div>,
      this.div,
    );
  }
}

export default Modal;
