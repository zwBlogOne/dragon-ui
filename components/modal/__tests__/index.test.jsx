import React from 'react';
import ReactDOM from 'react-dom';
import { render, shallow, mount } from 'enzyme';
import toJson from 'enzyme-to-json';
import Modal from '../index';

const { Header, Body, Footer } = Modal;

ReactDOM.createPortal = node => node;

describe('Modal', () => {
  it('renders basic Modal correctly', () => {
    const wrapper = render(
      <Modal>
        <Header title="标题" />
        <Body>我是一个模态框</Body>
        <Footer>页脚</Footer>
      </Modal>
    );
    expect(toJson(wrapper)).toMatchSnapshot();
  });

  it('renders customized Modal correctly', () => {
    const wrapper = render(
      <Modal
        width={400}
        radius
        visible
        animationType="fade"
        animationDuration={500}
      >
        <Header title="标题" />
        <Body>我是一个模态框</Body>
        <Footer>页脚</Footer>
      </Modal>
    );
    expect(toJson(wrapper)).toMatchSnapshot();
  });

  it('renders closable Modal correctly', () => {
    const wrapper = render(
      <Modal>
        <Header title="标题" onClose={() => {}} />
        <Body>我是一个模态框</Body>
        <Footer>页脚</Footer>
      </Modal>
    );
    expect(toJson(wrapper)).toMatchSnapshot();
  });

  it('triggers onClose callback correctly when clicks on closable icon', () => {
    const onClose = jest.fn();
    const wrapper = mount(
      <Modal>
        <Header title="标题" onClose={onClose} />
        <Body>我是一个模态框</Body>
        <Footer>页脚</Footer>
      </Modal>
    );
    wrapper.find('.ui-modal-close').simulate('click');
    expect(onClose).toHaveBeenCalled();
  });

  it('triggers onMaskClick callback correctly', () => {
    const onMaskClick = jest.fn();
    const wrapper = shallow(
      <Modal onMaskClick={onMaskClick}>
        <Header title="标题" />
        <Body>我是一个模态框</Body>
        <Footer>页脚</Footer>
      </Modal>
    );

    wrapper.find('.ui-modal').simulate('click');
    expect(onMaskClick).toHaveBeenCalled();
  });
});
