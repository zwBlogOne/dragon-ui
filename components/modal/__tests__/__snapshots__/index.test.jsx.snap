// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Modal renders basic Modal correctly 1`] = `
<div
  class="ui-modal"
  style="-webkit-animation-duration:300ms;-moz-animation-duration:300ms;-ms-animation-duration:300ms;-o-animation-duration:300ms;animation-duration:300ms;position:fixed;display:none"
>
  <div
    class="ui-modal-wrapper"
  >
    <div
      class="ui-modal-dialog zoom-leave"
      style="width:600px;min-width:270px;-webkit-animation-duration:300ms;-moz-animation-duration:300ms;-ms-animation-duration:300ms;-o-animation-duration:300ms;animation-duration:300ms"
    >
      <div
        class="ui-modal-header"
      >
        <div
          class="ui-modal-title"
        >
          标题
        </div>
        <div
          class="ui-modal-close"
        >
          <i
            class="ui-icon ui-icon-wrong theme-default"
          />
        </div>
      </div>
      <div
        class="ui-modal-body"
        style="height:auto"
      >
        我是一个模态框
      </div>
      <div
        class="ui-modal-footer"
      >
        页脚
      </div>
    </div>
  </div>
</div>
`;

exports[`Modal renders closable Modal correctly 1`] = `
<div
  class="ui-modal"
  style="-webkit-animation-duration:300ms;-moz-animation-duration:300ms;-ms-animation-duration:300ms;-o-animation-duration:300ms;animation-duration:300ms;position:fixed;display:none"
>
  <div
    class="ui-modal-wrapper"
  >
    <div
      class="ui-modal-dialog zoom-leave"
      style="width:600px;min-width:270px;-webkit-animation-duration:300ms;-moz-animation-duration:300ms;-ms-animation-duration:300ms;-o-animation-duration:300ms;animation-duration:300ms"
    >
      <div
        class="ui-modal-header"
      >
        <div
          class="ui-modal-title"
        >
          标题
        </div>
        <div
          class="ui-modal-close"
        >
          <i
            class="ui-icon ui-icon-wrong theme-default"
          />
        </div>
      </div>
      <div
        class="ui-modal-body"
        style="height:auto"
      >
        我是一个模态框
      </div>
      <div
        class="ui-modal-footer"
      >
        页脚
      </div>
    </div>
  </div>
</div>
`;

exports[`Modal renders customized Modal correctly 1`] = `
<div
  class="ui-modal radius fade-enter"
  style="-webkit-animation-duration:500ms;-moz-animation-duration:500ms;-ms-animation-duration:500ms;-o-animation-duration:500ms;animation-duration:500ms;position:fixed"
>
  <div
    class="ui-modal-wrapper"
  >
    <div
      class="ui-modal-dialog fade-enter"
      style="width:400px;min-width:270px;-webkit-animation-duration:500ms;-moz-animation-duration:500ms;-ms-animation-duration:500ms;-o-animation-duration:500ms;animation-duration:500ms"
    >
      <div
        class="ui-modal-header"
      >
        <div
          class="ui-modal-title"
        >
          标题
        </div>
        <div
          class="ui-modal-close"
        >
          <i
            class="ui-icon ui-icon-wrong theme-default"
          />
        </div>
      </div>
      <div
        class="ui-modal-body"
        style="height:auto"
      >
        我是一个模态框
      </div>
      <div
        class="ui-modal-footer"
      >
        页脚
      </div>
    </div>
  </div>
</div>
`;
